model/command_autocomplete.go

```diff
@@ -0,0 +1,398 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+package model
+
+import (
+	"encoding/json"
+	"net/url"
+	"reflect"
+
+	"github.com/pkg/errors"
+)
+
+// Argument types
+const (
+	TextInputArgumentType   = "TextInput"
+	StaticListArgumentType  = "StaticList"
+	DynamicListArgumentType = "DynamicList"
+)
+
+// AutocompleteData describes slash command autocomplete information.
+type AutocompleteData struct {
+	// Trigger of the command
+	Trigger string
+	// Text displayed to the user to help with the autocomplete
+	HelpText string
+	// Role of the user who should be able to see the autocomplete info of this command
+	RoleID string
+	// Arguments of the command. Arguments can be named or positional.
+	// If they are positional order in the list matters, if they are named order does not matter.
+	// All arguments should be either named or positional, no mixing allowed.
+	Arguments []*Argument
+	// Subcommands of the command
+	SubCommands []*AutocompleteData
+}
+
+// Argument describes an argument of the command. Arguments can be named or positional.
+// If Name is empty string Argument is positional otherwise it is named argument.
+// Named arguments are passed as --Name Argument_Value.
+type Argument struct {
+	// Name of the argument
+	Name string
+	// Text displayed to the user to help with the autocomplete
+	HelpText string
+	// Type of the argument
+	Type string
+	// Actual data of the argument(depends on the Type)
+	Data interface{}
+}
+
+// TextInputArgument describes text user can input as an argument.
+type TextInputArgument struct {
+	// Hint of the input text
+	Hint string
+	// Regex pattern to match
+	Pattern string
+}
+
+// StaticArgument describes an item in the StaticListArgument.
+type StaticArgument struct {
+	Item     string
+	HelpText string
+}
+
+// StaticListArgument is used to input one of the arguments from the list,
+// for example [yes, no], [true, false], and so on.
+type StaticListArgument struct {
+	PossibleArguments []StaticArgument
+}
+
+// DynamicListArgument is used when user wants to download possible argument list from the URL.
+type DynamicListArgument struct {
+	FetchURL string
+}
+
+// Suggestion describes single suggestion item sent to front
+type Suggestion struct {
```

Comment:
1/5 `s/Suggestion/AutocompleteSuggestion/` - `s Suggestion` is just not _suggestive_ enough? :)